src/session.ts

```diff
@@ -1,29 +1,77 @@
 import { Context } from './context'
 import { Middleware } from './types'
 
-export function session<SessionData extends object>({
-  ttl = Infinity,
-  store = new Map<string, { expires: number; session: SessionData }>(),
-  getSessionKey = (ctx: Context) =>
-    ctx.from && ctx.chat && `${ctx.from.id}:${ctx.chat.id}`,
-} = {}): Middleware.ExtFn<Context, { session?: SessionData }> {
-  const ttlMs = ttl * 1000
+export type StorageKeyFn<T extends Context> = (
+  ctx: T
+) => Promise<string | undefined>
 
+export interface Storage<T> {
+  getItem: (name: string) => Promise<T | undefined>
+  setItem: (name: string, value: T) => Promise<void>
+  deleteItem: (name: string) => Promise<void>
+}
+
+/**
+ * session middleware
+ *
+ * more 3rd-party session middlware
+ *
+ * see https://npm.im/search?q=telegraf-session
+ */
+export function session<SessionData extends object>(
+  makeKey: StorageKeyFn<Context> = makeDefaultKey,
+  storage: Storage<SessionData> = new MemorySessionStorage<SessionData>()
+): Middleware.ExtFn<Context, { session?: SessionData }> {
   return async (ctx, next) => {
-    const key = getSessionKey(ctx)
+    const key = await makeKey(ctx)
     if (key == null) {
       return await next(ctx)
     }
-    const now = Date.now()
-    const entry = store.get(key)
-    const ctx2 = Object.assign(ctx, {
-      session: entry == null || entry.expires < now ? undefined : entry.session,
-    })
+    const entry = await storage.getItem(key)
+    const ctx2 = Object.assign(ctx, { session: entry })
     await next(ctx2)
     if (ctx2.session == null) {
-      store.delete(key)
+      await storage.deleteItem(key)
     } else {
-      store.set(key, { session: ctx2.session, expires: now + ttlMs })
+      await storage.setItem(key, ctx2.session)
+    }
+  }
+}
+
+async function makeDefaultKey(ctx: Context): Promise<string | undefined> {
+  const fromId = ctx.from?.id
+  const chatId = ctx.chat?.id
+  if (fromId == null || chatId == null) {
+    return undefined
+  }
+  return `${fromId}:${chatId}`
+}
+
+export class MemorySessionStorage<T> implements Storage<T> {
+  private readonly ttl: number
+  private readonly store = new Map<string, { session: T; expires: number }>()
+
+  constructor(ttl = Infinity) {
+    this.ttl = ttl * 1000
+  }
```

Comment:
Is there a reason to put this into `MemorySessionStorage` only?

If you want to connect to MongoDB instead, it could still be very useful to store a `ttl`. So why not just keep it in the argument object of `session` and enable `ttl` for all storage types? :)

Reply:
It should be decided by storage,
e.q: redis built-in `expire` command

see https://redis.io/commands/expire

Reply:
Different storage has built-in TTL.

Reply:
I see, that makes sense.